src/ray/raylet/worker_killing_policy_group_by_owner.cc

```diff
@@ -0,0 +1,183 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/raylet/worker_killing_policy_group_by_owner.h"
+
+#include <gtest/gtest_prod.h>
+
+#include <boost/container_hash/hash.hpp>
+#include <unordered_map>
+
+#include "absl/container/flat_hash_map.h"
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/asio/periodical_runner.h"
+#include "ray/raylet/worker.h"
+#include "ray/raylet/worker_killing_policy.h"
+#include "ray/raylet/worker_pool.h"
+
+namespace ray {
+
+namespace raylet {
+
+GroupByOwnerIdWorkerKillingPolicy::GroupByOwnerIdWorkerKillingPolicy() {}
+
+const std::pair<std::shared_ptr<WorkerInterface>, bool>
+GroupByOwnerIdWorkerKillingPolicy::SelectWorkerToKill(
+    const std::vector<std::shared_ptr<WorkerInterface>> &workers,
+    const MemorySnapshot &system_memory) const {
+  if (workers.empty()) {
+    RAY_LOG_EVERY_MS(INFO, 5000) << "Worker list is empty. Nothing can be killed";
+    return std::make_pair(nullptr, /*should retry*/ false);
+  }
+
+  std::unordered_map<std::string, Group> group_map;
+  for (auto worker : workers) {
+    bool retriable = worker->GetAssignedTask().GetTaskSpecification().IsRetriable();
+    TaskID owner_id =
+        retriable ? worker->GetAssignedTask().GetTaskSpecification().ParentTaskId()
+                  : worker->GetAssignedTaskId();
+
+    auto it = group_map.find(owner_id.Binary());
+
+    if (it == group_map.end()) {
+      Group group(owner_id, retriable);
+      group.AddToGroup(worker);
+      group_map.emplace(owner_id.Binary(), std::move(group));
+    } else {
+      auto &group = it->second;
+      group.AddToGroup(worker);
+    }
+  }
+
+  std::vector<Group> sorted;
+  for (auto it = group_map.begin(); it != group_map.end(); ++it) {
+    sorted.push_back(it->second);
+  }
+
+  std::sort(
+      sorted.begin(), sorted.end(), [](Group const &left, Group const &right) -> bool {
+        int left_retriable = left.IsRetriable() ? 0 : 1;
+        int right_retriable = right.IsRetriable() ? 0 : 1;
+        if (left_retriable == right_retriable) {
+          return left.GetAssignedTaskTime() > right.GetAssignedTaskTime();
+        }
+        return left_retriable < right_retriable;
```

Comment:
Yea I think it is prioritizing retriable task:

The value is 0 if the task is retriable and is 1 otherwise.

The comp function here will return true when left < right, saying the left task (retriable) is "smaller" than the right (non-retriable), and therefore shows up first in the sorted result.

Reply:
Ah I see, I didn't see the line above that flips the flag; I was thinking 0 meant that the task was not retriable.

Maybe rename these vars to `left_no_retries` just for clarity?